### DISCUSSION TITLE
Allow array to return only valid items and not fail whole array

### DISCUSSION THREAD

**Discussion by bmeverett** [colinhacks/zod#1824] [Q&A] [answered]
When validating an array, I would like to only return the valid items in an array and not fail if one item fails. 

This is a very basic example
```ts
const schema = z.object({
  arr: z.array(z.number())
})

const result = schema.safeParse({arr: [1,2,"3"]})
```

I would like this to return only the valid items
``` ts
arr: [1,2]
```

I tried using `refine` on the array, but it seems to only be called when the array validates correctly. Is there a way to do this, or is this something you'd be willing to support?

**Comment by lucassarcanjo** [comment, marked as answer]
Hi @bmeverett,

I don't know if Zod maintainers will support this specific claim in the lib, but you can build a custom validator that uses `error.issues` to filter it manually based on invalid properties/items issued by zod.

I've implemented something like that just as a proof of concept:

```ts
const validator = (schema: z.Schema<any, any>) => (data: object) => {
  const response = schema.safeParse(data);

  if (response.success === false) {
    const errorPaths = response.error.issues.map((issue) => issue.path);

    let filteredData = data;
    errorPaths.forEach(
      (error) => (filteredData = filterByPath(filteredData, error))
    );

    return filteredData;
  }

  return data;
};

const result = validator(schema)(input);
```

The most difficult part is the `filterByPath()` function, which I've got something that works but maybe it's not perfect in performance terms and maybe does not cover all cases. 

I would suggest you develop your own for your needs. For reference, you can see my implementation [here in StackBlitz](https://stackblitz.com/edit/typescript-fxdfe5?file=index.ts).

**Comment by JacobWeisenburger** [comment]
@bmeverett 
Is this what you're looking for?

```ts
const schema = z.object( {
    arr: z.any().array().transform(
        arr => arr.map( item =>
            z.number().safeParse( item ).success ? item : undefined
        ).filter( Boolean )
    ),
} )

console.log( schema.parse( { arr: [ 1, 2, '3' ] } ) ) // { arr: [ 1, 2 ] }
```

**Comment by bmeverett** [comment]
Thanks @JacobWeisenburger and @lucassarcanjo I think I can work with these.

**Comment by alexgleason** [comment]
Here's what I ended up with, based on @JacobWeisenburger's answer:

```ts
function filteredArray<T>(schema: z.ZodType<T>) {
  return z.any().array().transform((arr) => (
    arr.map((item) => schema.safeParse(item).success ? item as T : undefined)
      .filter((item): item is T => Boolean(item))
  ));
}
```

Usage:

```ts
const noteSchema = z.object({
  tag: filteredArray(tagSchema).catch([]),
});
```

The type of `note.tag` is correct when used this way.

**Reply by AkisArou** [reply]
> Here's what I ended up with, based on @JacobWeisenburger's answer:
> 
> ```ts
> function filteredArray<T>(schema: z.ZodType<T>) {
>   return z.any().array().transform((arr) => (
>     arr.map((item) => schema.safeParse(item).success ? item as T : undefined)
>       .filter((item): item is T => Boolean(item))
>   ));
> }
> ```
> 
> Usage:
> 
> ```ts
> const noteSchema = z.object({
>   tag: filteredArray(tagSchema).catch([]),
> });
> ```
> 
> The type of `note.tag` is correct when used this way.

I modified your function a little bit, so any transformation and coercion applies.

```ts
function filteredArray<T>(schema: z.ZodType<T>) {
	return z
		.any()
		.array()
		.transform((arr) =>
			arr
				.map((item) => {
					const res = schema.safeParse(item);
					return res.success ? (res.data as T) : undefined;
				})
				.filter((item): item is T => Boolean(item)),
		);
}

```

**Comment by mysuf** [comment]
This is probably reason for me to leave Zod behind. Proposed solutions drop error messages or require to traverse paths. Basically, there is no way to make parsing more greedy.

I wonder why `safeParse` isn't implementing this. It's up to us developer how he handle the result. He can always throw provided error if result's data are insufficient.

My expectations from `safeParse`:
1) It should pass validation for all required (non-optional) fields and arrays with nonempty/min/max/length validator, otherwise returns `{ status: false, error }`
2) It should try to parse all optional fields and array entries, on invalid: logs issue, skips in case of array entry otherwise returns `default()`/`undefined`

From compatibility point of view it would require to change `ReturnType<safeParse>` to:
`{ status: false;  data: T | undefined; error: ZodError } | { success: true; data: T; }`

**Comment by d07RiV** [comment]
I'm new to zod and I'm trying to use it to replace my old validation which is very generous with parsing, so this array is something I definitely need. However I'm confused with these answers - why do we need safeParse? Can't we just do

    const NOTHING = Symbol.for('z-nothing')
    const z_safeArray = <T>(schema: z.ZodType<T | typeof NOTHING>) => z.array(schema.catch(NOTHING)).transform(arr => arr.filter((v): v is T => v !== NOTHING))

I'm using a symbol here in case `undefined` is a valid value in the schema. Am I missing something here that `safeParse` would cover better? Or is `.catch` simply newer than this discussion?

**Reply by BrendanC23** [reply]
That seems like a reasonable approach.

**Reply by mysuf** [reply]
This way, you loose information that there was an issue. It either parses it successfuly or throws an error. There is currently nothing between like a warning message. I just said that safeParse can be easily extended to support shades of grey. Actually year ago I forked this repo, made few changes and it works. So currently I use my fork instead of this one.
